From c9d2a3f9a0605ef66d2faeef38b75cad605a87a3 Mon Sep 17 00:00:00 2001
From: Zhihao Cheng <chengzhihao1@huawei.com>
Date: Wed, 14 Jun 2023 19:35:56 +0800
Subject: [PATCH 026/780] [Huawei] ext4: Limit retry times if ext4_writepages
 occurs ENOSPC

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188819

--------------------------------

Limit retry times to 3 when ext4_writepages occurs ENOSPC and there
are free clusters.

Signed-off-by: Zhihao Cheng <chengzhihao1@huawei.com>
---
 fs/ext4/inode.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index 388b13f37bd3..708cc36d2a37 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -2450,7 +2450,7 @@ static int mpage_map_one_extent(handle_t *handle, struct mpage_da_data *mpd)
  */
 static int mpage_map_and_submit_extent(handle_t *handle,
 				       struct mpage_da_data *mpd,
-				       bool *give_up_on_write)
+				       bool *give_up_on_write, int *retries)
 {
 	struct inode *inode = mpd->inode;
 	struct ext4_map_blocks *map = &mpd->map;
@@ -2478,7 +2478,8 @@ static int mpage_map_and_submit_extent(handle_t *handle,
 			 * is non-zero, a commit should free up blocks.
 			 */
 			if ((err == -ENOMEM) ||
-			    (err == -ENOSPC && EXT4_SB(sb)->s_mb_free_pending)) {
+			    (err == -ENOSPC && EXT4_SB(sb)->s_mb_free_pending &&
+			     ++(*retries) < 3)) {
 				if (progress)
 					goto update_disksize;
 				return err;
@@ -2695,6 +2696,7 @@ static int ext4_writepages(struct address_space *mapping,
 	struct ext4_sb_info *sbi = EXT4_SB(mapping->host->i_sb);
 	struct blk_plug plug;
 	bool give_up_on_write = false;
+	int __maybe_unused retries = 0;
 
 	if (unlikely(ext4_forced_shutdown(EXT4_SB(inode->i_sb))))
 		return -EIO;
@@ -2840,7 +2842,7 @@ static int ext4_writepages(struct address_space *mapping,
 		ret = mpage_prepare_extent_to_map(&mpd);
 		if (!ret && mpd.map.m_len)
 			ret = mpage_map_and_submit_extent(handle, &mpd,
-					&give_up_on_write);
+					&give_up_on_write, &retries);
 		/*
 		 * Caution: If the handle is synchronous,
 		 * ext4_journal_stop() can wait for transaction commit
-- 
2.17.1

